File: 297-Serialize and Deserialize Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
# time complexity: O(n)
# space complexity: O(n)
        def dfs(root):
            if not root:
                return "#,"
            string = ''
            string += str(root.val) + ','
            string += dfs(root.left)
            string += dfs(root.right)
            return string
        res = dfs(root)
        return res
        
    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
# time complexity: O(n)
# space complexity: O(n)
        def dfs(nodeLinst):
            if nodeList[0] == '#':
                nodeList.pop(0)
                return None
            root = TreeNode(nodeList[0])
            nodeList.pop(0)
            root.left = dfs(nodeList)
            root.right = dfs(nodeList)
            return root
        
        nodeList = data.split(',')[:-1]
        return dfs(nodeList)

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))

File: JavaScript/297-Serialize and Deserialize Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * Encodes a tree to a single string.
 *
 * @param {TreeNode} root
 * @return {string}
 */
var serialize = function(root) {
    let data = [];
    let dfs = (root) =>{
        if (!root){
            data.push('#');
            return;
        }
        data.push(root.val.toString())
        dfs(root.left)
        dfs(root.right)
    }
    dfs(root);
    return data.join(" ");
};

/**
 * Decodes your encoded data to tree.
 *
 * @param {string} data
 * @return {TreeNode}
 */
var deserialize = function(data) {
    data = data.split(" ");
    let dfs = () =>{
        if (data[0] === '#') return null;
        let root = new TreeNode(parseInt(data.shift()) )
        if (data[0]!=='#'){
            root.left = dfs()
        } else{
            data.shift()
        }
    
        if (data[0]!=='#'){
            root.right = dfs();
        } else{
            data.shift();
        }
        return root;
    }
    return dfs();
};